ffmpeg-2404-sdk/snapcraft.yaml

```
@@ -480,5 +480,5 @@ parts:
     after: [deps]
     override-prime: |
       craftctl default
-      sed -i '/usr\/lib\/x86_64-linux-gnu\/libvulkan.so.1\*/d' ${CRAFT_PART_SRC}/lists/*.list
+      sed -i '/usr\/lib\/${CRAFT_ARCH_TRIPLET_BUILD_FOR}\/libvulkan.so.1\*/d' ${CRAFT_PART_SRC}/lists/*.list
```

Best use some other sed delimiters so you don't have to escape. 

Also only mangle the relevant file.

```suggestion
      sed -i '|usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/libvulkan.so.1\*|d' ${CRAFT_PART_SRC}/lists/{CRAFT_ARCH_BUILD_FOR}.list
```

---

We're removing the complete line here. The `/` is not a delimiter, it belongs to the line itself, which will be removed. Also the list file names are `mesa-2404.amd64.list` and similar, so only `${CRAFT_ARCH_BUILD_FOR}` will not work. Rather, I think keeping it as `*` is a better option.

---

@Saviq

> Best use some other sed delimiters so you don't have to escape.

According to [the sed manual > sed addresses > selecting lines by text matching](https://www.gnu.org/software/sed/manual/html_node/Regexp-Addresses.html#Regexp-Addresses):

```txt
 ‘\%REGEXP%’
      (The ‘%’ may be replaced by any other single character.)

      This also matches the regular expression REGEXP, but allows one to
      use a different delimiter than ‘/’.  This is particularly useful if
      the REGEXP itself contains a lot of slashes, since it avoids the
      tedious escaping of every ‘/’.  If REGEXP itself includes any
      delimiter characters, each must be escaped by a backslash (‘\’).
```

an additional leading backslash will be required if you need to change the delimiter:

```bash
sed -i '\|usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/libvulkan.so.1\*|d' ${CRAFT_PART_SRC}/lists/{CRAFT_ARCH_BUILD_FOR}.list
```

However, this also won't work as [in Bash/Posix `sh` shell scripts ${parameter} expansion syntax in the single quote string will _not_ be expanded](https://www.gnu.org/software/bash/manual/html_node/Single-Quotes.html), so it should be changed to:

```bash
sed -i "\\|usr/lib/...
```

Assuming the `*` character in the matching pattern matches a literal asterisk symbol, the resulting command would be:

```bash
sed -i "\\|usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/libvulkan.so.1\\*|d" ${CRAFT_PART_SRC}/lists/${CRAFT_ARCH_BUILD_FOR}.list
```

@soumyaDghosh 

> The / is not a delimiter, it belongs to the line itself, which will be removed.

I agree with @Saviq here, the `d` sed command does remove the matched line however the backslashes at the start of the expression and the one before `d` are indeed delimiters of the sed's "selecting lines by text matching" syntax.  Replacing the delimiter with other symbols to avoid escaping forward slashes is a reasonable change.

> the list file names are mesa-2404.amd64.list and similar, so only ${CRAFT_ARCH_BUILD_FOR} will not work. Rather, I think keeping it as * is a better option.

I agree.

---

Commands that can reveal what the `sed` process will get from argv[] _after_ expansion:

```sh
$ CRAFT_ARCH_TRIPLET_BUILD_FOR=x86_64-linux-gnu
$ CRAFT_PART_SRC=/root/parts/cleanup/src
$ CRAFT_ARCH_BUILD_FOR=amd64
$ command_components=($(echo sed -i "\\|usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/libvulkan.so.1\\*|d" ${CRAFT_PART_SRC}/lists/${CRAFT_ARCH_BUILD_FOR}.list))
$ for component in "${command_components[@]}"; do
> echo "${component}"
> done
sed
-i
\|usr/lib/x86_64-linux-gnu/libvulkan.so.1\*|d
/root/parts/cleanup/src/lists/amd64.list
```

---

After simulating the operation in an LXD container the sed command _should_ be replaced to:

```bash
sed -i "\\|usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/libvulkan.so.1\\*|d" ${CRAFT_PART_SRC}/lists/*.${CRAFT_ARCH_BUILD_FOR}.list
```

Terminal output snippet:

```txt
root@ubuntu-2404:~# sed -i "\\|usr/lib/${CRAFT_ARCH_TRIPLET_BUILD_FOR}/libvulkan.so.1\\*|d" ${CRAFT_PART_SRC}/lists/*.${CRAFT_ARCH_BUILD_FOR}.list
root@ubuntu-2404:~# diff -u /root/parts/cleanup/src/lists/mesa-2404.amd64.list{.orig,}
--- /root/parts/cleanup/src/lists/mesa-2404.amd64.list.orig     2024-08-23 08:37:08.781069767 +0000
+++ /root/parts/cleanup/src/lists/mesa-2404.amd64.list  2024-08-23 08:38:54.304682738 +0000
@@ -173,7 +173,6 @@
 usr/lib/x86_64-linux-gnu/libva-x11.so.2*
 usr/lib/x86_64-linux-gnu/libva.so.2*
 usr/lib/x86_64-linux-gnu/libvdpau.so.1*
-usr/lib/x86_64-linux-gnu/libvulkan.so.1*
 usr/lib/x86_64-linux-gnu/libvulkan_intel.so
 usr/lib/x86_64-linux-gnu/libvulkan_intel_hasvk.so
 usr/lib/x86_64-linux-gnu/libvulkan_lvp.so
```

---

The initial and final `/` are delimiters (`/foo\/bar/d` == `|foo/bar|d`), it's just easier to read without all the `\/`.

`*.${CRAFT_ARCH_BUILD_FOR}.list` would work, too. Anyway, your call. The approach is sane in general.

---

Thanks @brlin-tw for the sleuthing. Didn't realize `sed` wants the initial backslash - since it's not the case for `s|foo|bar|`…